### ISSUE TITLE
wandb.log is only logging metrics from one call and not logging metrics from other log calls in one epoch

### ISSUE THREAD

**Issue by Breakpoint23** [wandb/wandb#7021]
### Current Behavior

I'm using wandb to watch my pytorch model training and I'm calling the log function every 50 batch during training and every 50 batch in testing. And then after the training and testing is done I'm calling it at the end of the epoch to log avg metrics for the epoch. But All i'm seeing in the App is the testing log metrics logged while training(first log call in the epoch). I'm not able to access the metrics from other log calls (in testing and after the epoch).

### Expected Behavior

What I expect is that it logs the metrics when ever I call the log function

### Steps To Reproduce

`def train_epoch(model, train_loader, optimizer, triplet_loss,epoch):
    model.train()
    device=torch.device('cuda')
    model.to(device)
    total_loss = []
    avg_loss= 0
    exp_ct=0
    start_time=time.time()
    for  i,(X_a,X_p,X_n) in tqdm(enumerate(train_loader),total=len(train_loader)):
        X_a,X_p,X_n=X_a.to(device),X_p.to(device),X_n.to(device)
        optimizer.zero_grad()
        anchor = model(X_a)
        positive = model(X_p)
        negative = model(X_n)
        loss = triplet_loss(anchor, positive, negative)
        loss.backward()
        optimizer.step()
        total_loss += [loss.item()]
        avg_loss += loss.item()
        exp_ct+=len(anchor)
        
        
        if i%50==0:
            wandb.log({"training loss":loss.item()},step=exp_ct)
    avg_loss /=len(train_loader)
    
    return total_loss,avg_loss

def test_epoch(model, test_loader, triplet_loss):
    model.eval()
    device=torch.device('cuda')
    model.to(device)
    total_loss = []
    eu_dist_pos=[]
    avg_eu_dist_pos=0
    eu_dist_neg=[]
    avg_eu_dist_neg=0
    avg_loss=0
    exp_ct=0
    flag=True
    with torch.no_grad():
        for i,(X_a,X_p,X_n) in tqdm(enumerate(test_loader),total=len(test_loader)):
            X_a,X_p,X_n=X_a.to(device),X_p.to(device),X_n.to(device)
            anchor = model(X_a)
            positive = model(X_p)
            negative = model(X_n)
            loss = triplet_loss(anchor, positive, negative)
            total_loss += [loss.item()]
            avg_loss += loss.item()
            
            pos_dist= (anchor - positive).pow(2).sum(axis=1).sqrt()
            eu_dist_pos+=pos_dist.tolist()
            avg_eu_dist_pos+=pos_dist.mean().item()
            
            neg_dist= (anchor - negative).pow(2).sum(axis=1).sqrt()
            eu_dist_neg+=neg_dist.tolist()
            avg_eu_dist_neg+=neg_dist.mean().item()
            
            exp_ct+=len(anchor)
            
            if i%100==0:
                wandb.log({"testing loss":loss.item(),"Pos Distance":pos_dist.mean().item(),"Neg Distance":neg_dist.mean().item()},step=exp_ct)                
        avg_loss/=len(test_loader)
        avg_eu_dist_pos/=len(test_loader)
        avg_eu_dist_neg/=len(test_loader)
    
    return total_loss,avg_loss,avg_eu_dist_pos,avg_eu_dist_neg`

### Environment

OS: Linux 

Browsers: Waterfox

Version: 0.16.2/0.16.3

**Comment by Breakpoint23** [comment]
Following is my training script  

`from torch.optim import Adam
  from sklearn.model_selection import train_test_split
  from tqdm import tqdm
  
  train_loader=DataLoader(train_dataset,batch_size=128,shuffle=True)
  test_loader=DataLoader(test_dataset,batch_size=128,shuffle=True)
  model=CNN_LSTM(5,30,128)
  criterion=nn.TripletMarginLoss()
  optimizer=Adam(model.parameters(),lr=0.0003)
  device=torch.device('cuda')
  epochs=50
  
  config = dict(
      epochs=50,
      classes=10,
      features=5,
      window=30,
      batch_size=128,
      hidden=128,
      learning_rate=0.0003,
      dataset="Keystroke3",
      architecture="CNN_LSTM1")
  
  with wandb.init(project="Keystroke1",config=config):
      wandb.config=config
      wandb.watch(model,criterion,log='all',log_freq=25)
      for epoch in range(epochs):
          print(f'---------------------------------------------epoch {epoch}--------------------------------------------------')
  
          train_loss,avg_train_loss = train_epoch(model, train_loader, optimizer, criterion,epoch)
          test_loss,avg_test_loss,avg_pos_dist,avg_neg_dist = test_epoch(model,test_loader,criterion)
  
          print(f"Epoch {epoch+1}, Avg. Train Loss: {avg_train_loss:.4f},Avg. Test Loss: {avg_test_loss:.4f},Avg. Pos distance:{avg_pos_dist:.4f},Avg. Neg distance:{avg_neg_dist:.4f}")
          wandb.log({"Train_loss":avg_train_loss,"Test_loss":avg_test_loss,"Avg. Distance(P)":avg_pos_dist,"Avg. Distance(N)":avg_neg_dist,"epoch":epoch},step=epoch)
          torch.save(model.state_dict(),f"model_{epoch}.h5")
          torch.onnx.export(model,torch.rand(1,5,30).to(device),f"model_{epoch}.onnx")
          wandb.save(f"model_{epoch}.onnx")
`

**Comment by Breakpoint23** [comment]
> Hi how did you fix this?

@smandava98  It turns out the the step (Argument in the wand.log )counter needs to be consistent and you can not have different step counters inside one run. 
- Here I had one step counter for training and the other for testing, So the error was the the step counter should always increase but once I move from training to testing it would decrease (because I have less test batches than training batches ).
- I could not figure that out before because there was no error message printed in the logs. It did not notify me that this is an issue.
- For separate counters I suggest instead of using step argument just log the train/test step counter as a metric and then update the graphs in  the run to use that. (wandb.log({"training loss": some number,"train_log_cnt": some number}))

Hope this helps.
